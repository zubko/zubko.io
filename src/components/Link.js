/**
 * Link with styling
 *
 * @flow
 */

import * as React from 'react';
import { Link as GatsbyLink } from 'gatsby';

import { Styles } from '../Theme';

export default ({
  useAnchor,
  usePlainStyle,
  children,
  to,
  ...otherProps
}: Object) => {
  const internal = to === '#' || to.indexOf('/') === 0;
  const style = usePlainStyle ? Styles.linkPlain : Styles.linkAnimated;
  if (!useAnchor && internal) {
    return (
      <GatsbyLink to={to} css={style} {...otherProps}>
        {children}
      </GatsbyLink>
    );
  }
  const sameTab = to.indexOf('mailto:') === 0;
  return (
    <a
      href={to}
      css={style}
      {...(!sameTab ? { target: '_blank' } : null)}
      {...otherProps}>
      {children}
    </a>
  );
};
