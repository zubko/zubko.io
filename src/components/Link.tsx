/**
 * Link with styling
 */

import { Link as GatsbyLink } from "gatsby";
import * as React from "react";

import { Styles } from "../Theme";

type Props = {
  useAnchor?: boolean;
  usePlainStyle?: boolean;
  children?: React.ReactNode;
  to: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void;
};

export const Link = ({
  useAnchor,
  usePlainStyle,
  children,
  to,
  ...otherProps
}: Props) => {
  const internal = to === "#" || to.indexOf("/") === 0;
  const style = usePlainStyle ? Styles.linkPlain : Styles.linkAnimated;
  if (!useAnchor && internal) {
    return (
      <GatsbyLink to={to} css={style} {...otherProps}>
        {children}
      </GatsbyLink>
    );
  }
  const sameTab = to.indexOf("mailto:") === 0;
  return (
    <a
      href={to}
      css={style}
      {...(!sameTab ? { target: "_blank", rel: "noopener noreferrer" } : null)}
      {...otherProps}
    >
      {children}
    </a>
  );
};
