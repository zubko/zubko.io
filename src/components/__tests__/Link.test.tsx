import React from "react";
import renderer from "react-test-renderer";

import Link from "../Link";

describe("Link", () => {
  it("inner animated link renders without regressions", () => {
    const tree = renderer.create(<Link to="/inner-link">Title</Link>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("inner plain link renders without regressions", () => {
    const tree = renderer
      .create(
        <Link to="/inner-link" usePlainStyle>
          Title
        </Link>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("outer link renders without regressions", () => {
    const tree = renderer
      .create(<Link to="https://outer.link.to">Title</Link>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
