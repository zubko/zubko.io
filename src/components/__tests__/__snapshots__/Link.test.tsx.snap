// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link inner animated link renders without regressions 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-1k30ml3{color:#2C8127;-webkit-text-decoration:none;text-decoration:none;background-image:linear-gradient(
    to bottom,
    currentColor 0%,
    currentColor 100%
  );-webkit-background-position:0 100%;background-position:0 100%;background-repeat:repeat-x;-webkit-background-size:4px 1px;background-size:4px 1px;-webkit-transition:background-size 0s;transition:background-size 0s;}.css-1k30ml3:hover{color:black;-webkit-text-decoration:none;text-decoration:none;background-image:linear-gradient(
      to bottom,
      #bcebb9 0%,
      #bcebb9 100%
    );-webkit-background-size:4px 100%;background-size:4px 100%;-webkit-transition:background-size 0.15s;transition:background-size 0.15s;}",
      }
    }
    data-emotion="css 1k30ml3"
  />,
  <a
    className="css-1k30ml3"
    href="/inner-link"
  >
    Title
  </a>,
]
`;

exports[`Link inner plain link renders without regressions 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-gpp53d{color:#2C8127;-webkit-text-decoration:none;text-decoration:none;background-image:linear-gradient(
    to top,
    rgba(0, 0, 0, 0),
    rgba(0, 0, 0, 0) 1px,
    currentColor 1px,
    currentColor 2px,
    rgba(0, 0, 0, 0) 2px
  );}.css-gpp53d:hover{background-image:none;-webkit-filter:brightness(1.3);filter:brightness(1.3);}",
      }
    }
    data-emotion="css gpp53d"
  />,
  <a
    className="css-gpp53d"
    href="/inner-link"
  >
    Title
  </a>,
]
`;

exports[`Link outer link renders without regressions 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-1k30ml3{color:#2C8127;-webkit-text-decoration:none;text-decoration:none;background-image:linear-gradient(
    to bottom,
    currentColor 0%,
    currentColor 100%
  );-webkit-background-position:0 100%;background-position:0 100%;background-repeat:repeat-x;-webkit-background-size:4px 1px;background-size:4px 1px;-webkit-transition:background-size 0s;transition:background-size 0s;}.css-1k30ml3:hover{color:black;-webkit-text-decoration:none;text-decoration:none;background-image:linear-gradient(
      to bottom,
      #bcebb9 0%,
      #bcebb9 100%
    );-webkit-background-size:4px 100%;background-size:4px 100%;-webkit-transition:background-size 0.15s;transition:background-size 0.15s;}",
      }
    }
    data-emotion="css 1k30ml3"
  />,
  <a
    className="css-1k30ml3"
    href="https://outer.link.to"
    rel="noopener noreferrer"
    target="_blank"
  >
    Title
  </a>,
]
`;
